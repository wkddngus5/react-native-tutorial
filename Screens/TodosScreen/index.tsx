import React from 'react';
import { View, Text, Button } from 'react-native';
import { BottomTabNavigationProp } from '@react-navigation/bottom-tabs';
import { RootStackParamList } from '../RootStackParamList';

type TodosScreenNavigationProp = BottomTabNavigationProp<
  RootStackParamList,
  'Todos'
>;

export default function TodosScreen({
  navigation,
}: {
  navigation: TodosScreenNavigationProp;
}) {
  return (
    <View style={{ flex: 1, alignItems: 'center', justifyContent: 'center' }}>
      <Text>Todos Screen</Text>
      <Button
        title="Go to Home"
        onPress={() => navigation.navigate('Statistics')}
      />
      <Button title="Go to back" onPress={() => navigation.goBack()} />
    </View>
  );
}
