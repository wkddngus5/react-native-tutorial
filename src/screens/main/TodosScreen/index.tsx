import React, { useCallback, useMemo, useState } from 'react';
import { View, Text, Button, TextInput, Modal, StyleSheet } from 'react-native';
import { RootStoreState } from '../../../reducer/index';
import { addTodo, toggleDoneTodo } from '../../../actions/Todo';
import { useDispatch, useSelector } from 'react-redux';
import Todo from '../../../../classes/Todo';
import TodoComponent from '../../../components/Todo';

export default function TodosScreen() {
  const dispatch = useDispatch();
  const [showModal, setShowModal] = useState<boolean>(false);

  const todos = useSelector<RootStoreState, Todo[]>((store): Todo[] => {
    return store.todo.todos;
  });
  const [title, setTitle] = useState<string>('');

  const onAddTodo = useCallback(() => {
    dispatch(addTodo({ title }));
    setTitle('');
    setShowModal(true);
  }, [dispatch, title]);

  const todoList = useMemo(() => {
    return todos.map((todo, index) => {
      const onToggleDoneTodo = () => {
        dispatch(toggleDoneTodo({ index, isDone: todo.isDone }));
      };
      return (
        <TodoComponent
          todo={todo}
          key={index}
          onChangeDone={onToggleDoneTodo}
        />
      );
    });
  }, [todos, dispatch]);

  return (
    <View style={{ flex: 1, alignItems: 'center', justifyContent: 'center' }}>
      <Modal visible={showModal}>
        <View style={styles.modalBody}>
          <Text style={styles.modalText}>new Todo is added!</Text>
          <Button title="OK" onPress={() => setShowModal(false)} />
        </View>
      </Modal>
      <Text>Todos Screen</Text>
      <TextInput
        value={title}
        onChangeText={setTitle}
        placeholder="todo title"
      />
      <Button title="ADD" onPress={onAddTodo} />
      {todoList}
    </View>
  );
}

const styles = StyleSheet.create({
  modalBody: {
    justifyContent: 'center',
    alignContent: 'center',
    height: '100%',
  },
  modalText: {
    textAlign: 'center',
  },
});
