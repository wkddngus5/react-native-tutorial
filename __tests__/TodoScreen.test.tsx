import React from 'react';
import { fireEvent, render } from '@testing-library/react-native';
import configureStore from 'redux-mock-store';
import TodosScreen from '../src/screens/main/TodosScreen';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import rootReducer from '../src/reducer';
import { ReactTestInstance } from 'react-test-renderer';

const mockStore = configureStore([]);

describe('Todo Component', () => {
  it('render', () => {
    const store = mockStore({
      todo: { todos: [] },
    });
    const { queryByText, toJSON } = render(
      <Provider store={store}>
        <TodosScreen />
      </Provider>,
    );
    expect(queryByText('Todos Screen')).toBeTruthy();
    expect(toJSON()).toMatchSnapshot();
  });
  it('add todo', () => {
    const { getByPlaceholderText, queryByText } = render(
      <Provider store={createStore(rootReducer)}>
        <TodosScreen />
      </Provider>,
    );
    const $inputText = getByPlaceholderText('todo title');
    fireEvent.changeText($inputText, 'new todo');
    fireEvent.press(queryByText('ADD') as ReactTestInstance);
    expect(queryByText('new todo')).toBeTruthy();
  });
});
