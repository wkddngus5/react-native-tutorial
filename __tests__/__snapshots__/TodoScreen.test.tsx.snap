// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo Component render 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text>
    Todos Screen
  </Text>
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    placeholder="todo title"
    rejectResponderTermination={true}
    underlineColorAndroid="transparent"
    value=""
  />
  <View
    accessibilityRole="button"
    accessibilityState={Object {}}
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        ADD
      </Text>
    </View>
  </View>
</View>
`;
